File: 5.12/main.c
#include "../nob.h"

typedef struct {
	ll left;
	ll right;
} Range;

typedef struct {
    Range *items;
    size_t count;
    size_t capacity;
} Ranges;

typedef struct {
    ll *items;
    size_t count;
    size_t capacity;
} LLDA;

int compare(const void *a, const void *b)
{
    ll x = *(const ll *)a;
    ll y = *(const ll *)b;

    if (x < y) return -1;
    if (x > y) return 1;
    return 0;
}

int compare_ranges(const void *a, const void *b) {
    const Range *r1 = a;
    const Range *r2 = b;
    if (r1->left < r2->left) return -1;
    if (r1->left > r2->left) return 1;
    return 0;
}

size_t is_fresh(Range *arr, size_t len, ll id) {
	ssize_t l = 0, r = len-1;

	while(l <= r) {
		size_t mid = l + (r-l) / 2;
		if(arr[mid].left <= id && id <= arr[mid].right) {
			return 1;
		} else if(id < arr[mid].left) {
			r = mid - 1;
		} else {
			l = mid + 1;	
		}
	}

	return 0;
}

int main(int argc, char **argv) {

    FILE *input = open_input_or_die(argc, argv);
    

	char *line = NULL;
  	size_t len = 0;
  	size_t change = 0; // 0 ranges | 1 elements 

	Ranges ranges = {0};
	LLDA elements   = {0};
  	
   	while (getline(&line, &len, input) != -1) {
   		if(strlen(line) == 1){
   			 change = 1;
   			 continue;
   		}
   		
 		line[strcspn(line, "\r\n")] = '\0';

 		if(change == 0){
 			Range r = {0};
 	
			if (sscanf(line, "%lld-%lld", &r.left, &r.right) != 2) {
	            fprintf(stderr, "Warning: malformed range '%s'.\n", line);
	            exit(1);
	        }
		
 			da_append(&ranges, r);
 		}else {
 			da_append(&elements, atoll(line));
 		} 		 
   	}

    fclose(input);

    ll part1 = 0;
    ll part2 = 0;

    qsort(elements.items, elements.count, sizeof(ll), compare);
	qsort(ranges.items, ranges.count, sizeof(Range), compare_ranges);

	int idx = 0;
	Ranges merged = {0};

	da_append(&merged, ranges.items[0]);

	for(size_t i = 1; i < ranges.count; ++i) {
		Range range = ranges.items[i];
		if(range.left <= merged.items[idx].right + 1) {
			if(range.right > merged.items[idx].right) merged.items[idx].right = range.right;
		} else {
			idx++;
			da_append(&merged, range);
		}
	}
	
	for(size_t i = 0; i < elements.count; ++i) {
		part1 += is_fresh(merged.items, merged.count, elements.items[i]);		
	}

	for(size_t i = 0; i < merged.count; ++i) {
		part2 += (merged.items[i].right - merged.items[i].left) + 1;		
	}

	printf("part1: %lld\n", part1);
	printf("part2: %lld\n", part2);
    

    free(line);
    da_free(&ranges);
    da_free(&elements);
    da_free(&merged);

    return 0;
}

File: 5.12/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead, BufReader};

#[derive(Debug, Clone)]
struct Range {
    left: i64,
    right: i64,
}

fn merge_ranges(ranges: &mut Vec<Range>) -> Vec<Range> {
    if ranges.is_empty() {
        return Vec::new();
    }

    ranges.sort_by_key(|r| r.left);
    let mut merged = vec![ranges[0].clone()];

    for r in ranges.iter().skip(1) {
        let last = merged.last_mut().unwrap();
        if r.left <= last.right + 1 {
            last.right = last.right.max(r.right);
        } else {
            merged.push(r.clone());
        }
    }

    merged
}

fn is_fresh(ranges: &[Range], id: i64) -> bool {
    let mut left = 0;
    let mut right = ranges.len() as isize - 1;

    while left <= right {
        let mid = left + (right - left) / 2;
        let mid_range = &ranges[mid as usize];
        if mid_range.left <= id && id <= mid_range.right {
            return true;
        } else if id < mid_range.left {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    false
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        std::process::exit(1);
    }

    let file = File::open(&args[1])?;
    let reader = BufReader::new(file);

    let mut ranges: Vec<Range> = Vec::new();
    let mut elements: Vec<i64> = Vec::new();
    let mut reading_elements = false;

    for line_res in reader.lines() {
        let line = line_res?;
        let line = line.trim();

        if line.is_empty() {
            reading_elements = true;
            continue;
        }

        if reading_elements {
            elements.push(line.parse().expect("Invalid number"));
        } else {
            let parts: Vec<&str> = line.split('-').collect();
            if parts.len() != 2 {
                eprintln!("Malformed range '{}'", line);
                std::process::exit(1);
            }
            let left = parts[0].parse().expect("Invalid left number");
            let right = parts[1].parse().expect("Invalid right number");
            ranges.push(Range { left, right });
        }
    }

    let merged = merge_ranges(&mut ranges);
    elements.sort_unstable();

    let part1 = elements.iter().filter(|&&e| is_fresh(&merged, e)).count();
    let part2: i64 = merged.iter().map(|r| r.right - r.left + 1).sum();

    println!("part1: {}", part1);
    println!("part2: {}", part2);

    Ok(())
}
